File: scripts/update_jogos.py
import os
import json
import math
from datetime import datetime, timedelta, timezone

import requests

API_KEY = os.getenv("API_FOOTBALL_KEY", "").strip()
BASE_URL = "https://v3.football.api-sports.io"

TZ_OFFSET_HOURS = -3  # BRT
TZ = timezone(timedelta(hours=TZ_OFFSET_HOURS))

# Para API FREE: limita quantos jogos ganham probabilidade por atualização
MAX_PROB_GAMES = 30

# Para prob: últimos 3 jogos finalizados
LAST_N = 3


def api_get(path: str, params: dict):
    if not API_KEY:
        raise RuntimeError("API_FOOTBALL_KEY não configurada (GitHub Secrets).")
    headers = {"x-apisports-key": API_KEY}
    r = requests.get(f"{BASE_URL}{path}", headers=headers, params=params, timeout=35)
    r.raise_for_status()
    return r.json()


def now_brt():
    return datetime.now(TZ)


def iso_brt(dt: datetime) -> str:
    return dt.astimezone(TZ).strftime("%Y-%m-%d %H:%M BRT")


def is_live(status_short: str) -> bool:
    return status_short in {"1H", "HT", "2H", "ET", "P", "BT"}


def poisson_cdf(k: int, lam: float) -> float:
    if lam <= 0:
        return 1.0 if k >= 0 else 0.0
    s = 0.0
    for i in range(0, k + 1):
        s += math.exp(-lam) * (lam ** i) / math.factorial(i)
    return s


def prob_over(threshold: float, lam_total: float) -> float:
    if threshold == 1.5:
        return 1.0 - poisson_cdf(1, lam_total)
    if threshold == 2.5:
        return 1.0 - poisson_cdf(2, lam_total)
    return 0.0


def prob_under(threshold: float, lam_total: float) -> float:
    if threshold == 1.5:
        return poisson_cdf(1, lam_total)
    if threshold == 2.5:
        return poisson_cdf(2, lam_total)
    return 0.0


def fair_odd(p: float) -> float:
    if p <= 0:
        return 999.0
    return 1.0 / p


def safe_round_odd(x: float) -> float:
    return round(x + 1e-9, 2)


def team_lastN_stats(team_id: int, season: int) -> dict:
    data = api_get("/fixtures", {
        "team": team_id,
        "season": season,
        "last": LAST_N,
        "status": "FT-AET-PEN",
        "timezone": "America/Sao_Paulo",
    })
    fixtures = data.get("response", [])
    if not fixtures:
        return {"gf": 0.0, "ga": 0.0, "n": 0}

    gf = 0
    ga = 0
    n = 0
    for fx in fixtures:
        teams = fx.get("teams", {})
        goals = fx.get("goals", {})

        home = teams.get("home", {})
        away = teams.get("away", {})
        gh = goals.get("home")
        ga_ = goals.get("away")

        if gh is None or ga_ is None:
            continue

        if home.get("id") == team_id:
            gf += int(gh)
            ga += int(ga_)
            n += 1
        elif away.get("id") == team_id:
            gf += int(ga_)
            ga += int(gh)
            n += 1

    if n == 0:
        return {"gf": 0.0, "ga": 0.0, "n": 0}

    return {"gf": gf / n, "ga": ga / n, "n": n}


def estimate_probs(home_stats: dict, away_stats: dict) -> dict:
    if home_stats.get("n", 0) == 0 or away_stats.get("n", 0) == 0:
        return {}

    lam_home = (home_stats["gf"] + away_stats["ga"]) / 2.0
    lam_away = (away_stats["gf"] + home_stats["ga"]) / 2.0
    lam_total = max(0.05, lam_home + lam_away)

    p_o15 = prob_over(1.5, lam_total)
    p_o25 = prob_over(2.5, lam_total)
    p_u15 = prob_under(1.5, lam_total)
    p_u25 = prob_under(2.5, lam_total)

    return {
        "over15": {"p": round(p_o15 * 100, 1), "odd": safe_round_odd(fair_odd(p_o15))},
        "over25": {"p": round(p_o25 * 100, 1), "odd": safe_round_odd(fair_odd(p_o25))},
        "under15": {"p": round(p_u15 * 100, 1), "odd": safe_round_odd(fair_odd(p_u15))},
        "under25": {"p": round(p_u25 * 100, 1), "odd": safe_round_odd(fair_odd(p_u25))},
        "model": f"last{LAST_N}_poisson",
        "lam_total": round(lam_total, 2),
    }


def parse_fixture_datetime(fx) -> datetime | None:
    dt_str = fx.get("fixture", {}).get("date")
    if not dt_str:
        return None
    # ISO com Z ou offset
    dt = datetime.fromisoformat(dt_str.replace("Z", "+00:00")).astimezone(TZ)
    return dt


def main():
    now = now_brt()
    today = now.strftime("%Y-%m-%d")
    tomorrow = (now + timedelta(days=1)).strftime("%Y-%m-%d")

    # Busca hoje e amanhã (timezone SP)
    data_today = api_get("/fixtures", {"date": today, "timezone": "America/Sao_Paulo"})
    data_tom = api_get("/fixtures", {"date": tomorrow, "timezone": "America/Sao_Paulo"})
    fixtures = (data_today.get("response", []) or []) + (data_tom.get("response", []) or [])

    # Busca ao vivo separado (garante)
    live_data = api_get("/fixtures", {"live": "all", "timezone": "America/Sao_Paulo"})
    live_fixtures = live_data.get("response", []) or []
    live_ids = {fx.get("fixture", {}).get("id") for fx in live_fixtures if fx.get("fixture", {}).get("id")}

    games = []
    for fx in fixtures:
        fixture = fx.get("fixture", {})
        teams = fx.get("teams", {})
        league = fx.get("league", {})
        goals = fx.get("goals", {})

        fid = fixture.get("id")
        dt = parse_fixture_datetime(fx)
        if not fid or not dt:
            continue

        status_short = fixture.get("status", {}).get("short", "")
        live = (fid in live_ids) or is_live(status_short)

        home = teams.get("home", {})
        away = teams.get("away", {})

        item = {
            "id": fid,
            "ts": int(dt.timestamp()),
            "date": dt.strftime("%Y-%m-%d"),
            "time": dt.strftime("%H:%M"),
            "status": status_short,
            "live": bool(live),
            "home": home.get("name", ""),
            "away": away.get("name", ""),
            "homeId": home.get("id"),
            "awayId": away.get("id"),
            "league": league.get("name", ""),
            "leagueId": league.get("id"),
            "country": league.get("country", ""),
            "season": league.get("season"),
        }

        # guarda score (UI decide mostrar só no ao vivo)
        gh = goals.get("home")
        ga_ = goals.get("away")
        if gh is not None and ga_ is not None:
            item["score"] = {"home": gh, "away": ga_}

        games.append(item)

    # Ordena: ao vivo primeiro, depois por horário
    games.sort(key=lambda g: (0 if g["live"] else 1, g["ts"]))

    # Probabilidades limitadas
    team_cache = {}
    prob_done = 0
    for g in games:
        if prob_done >= MAX_PROB_GAMES:
            break
        home_id = g.get("homeId")
        away_id = g.get("awayId")
        season = g.get("season")
        if not home_id or not away_id or not season:
            continue

        key_h = (home_id, season)
        key_a = (away_id, season)
        if key_h not in team_cache:
            team_cache[key_h] = team_lastN_stats(home_id, season)
        if key_a not in team_cache:
            team_cache[key_a] = team_lastN_stats(away_id, season)

        probs = estimate_probs(team_cache[key_h], team_cache[key_a])
        if probs:
            g["probs"] = probs
            prob_done += 1

    out = {
        "updatedAt": iso_brt(now),
        "source": "API-Football",
        "counts": {
            "games": len(games),
            "probGames": prob_done,
            "maxProbGames": MAX_PROB_GAMES,
        },
        "games": games,
    }

    with open("jogos.json", "w", encoding="utf-8") as f:
        json.dump(out, f, ensure_ascii=False, indent=2)


if __name__ == "__main__":
    main()

File: scripts/update-jogos.js
/**
 * Gera jogos.json (BRT) com:
 * - fixtures de HOJE e AMANHÃ (BRT)
 * - jogos AO VIVO (live=all)
 * - kickoffTs (timestamp), placar quando ao vivo e status
 *
 * Requer Secret: API_FOOTBALL_KEY
 */

const fs = require("fs");
const path = require("path");

const TZ = "America/Sao_Paulo";
const API_BASE = "https://v3.football.api-sports.io";
const API_KEY = process.env.API_FOOTBALL_KEY || "";

// ✅ Ajuste de sanidade: depois de X minutos do kickoff, não pode continuar "ao vivo"
const LIVE_STALE_MINUTES = 140; // ~2h20

if (!API_KEY) {
  console.error("❌ Missing API_FOOTBALL_KEY. Configure em Settings > Secrets > Actions.");
  process.exit(1);
}

function brtNowParts() {
  const fmt = new Intl.DateTimeFormat("en-CA", {
    timeZone: TZ,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    hour: "2-digit",
    minute: "2-digit",
    second: "2-digit",
    hour12: false,
  });

  const parts = fmt.formatToParts(new Date()).reduce((acc, p) => {
    acc[p.type] = p.value;
    return acc;
  }, {});

  const date = `${parts.year}-${parts.month}-${parts.day}`;
  const time = `${parts.hour}:${parts.minute}:${parts.second}`;
  return { date, time };
}

function addDaysYYYYMMDD(dateStr, days) {
  const [y, m, d] = dateStr.split("-").map(Number);
  const base = new Date(Date.UTC(y, m - 1, d, 12, 0, 0));
  base.setUTCDate(base.getUTCDate() + days);
  const yy = base.getUTCFullYear();
  const mm = String(base.getUTCMonth() + 1).padStart(2, "0");
  const dd = String(base.getUTCDate()).padStart(2, "0");
  return `${yy}-${mm}-${dd}`;
}

function formatBRTFromTimestampSec(tsSec) {
  const dt = new Date(tsSec * 1000);

  const fmtDate = new Intl.DateTimeFormat("en-CA", {
    timeZone: TZ,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
  }).format(dt);

  const fmtTime = new Intl.DateTimeFormat("pt-BR", {
    timeZone: TZ,
    hour: "2-digit",
    minute: "2-digit",
    hour12: false,
  }).format(dt);

  return { date: fmtDate, time: fmtTime };
}

async function apiGet(pathname, qs = {}) {
  const url = new URL(API_BASE + pathname);
  for (const [k, v] of Object.entries(qs)) url.searchParams.set(k, String(v));

  for (let attempt = 1; attempt <= 3; attempt++) {
    const res = await fetch(url.toString(), {
      headers: { "x-apisports-key": API_KEY },
    });

    if (res.ok) return res.json();

    const txt = await res.text().catch(() => "");
    console.warn(`⚠️ API ${res.status} attempt ${attempt}: ${txt.slice(0, 180)}`);

    if (attempt < 3) await new Promise((r) => setTimeout(r, 900 * attempt));
  }

  throw new Error("API request failed after retries.");
}

function normalizeStatusShort(short) {
  if (!short) return "UNK";
  return String(short).toUpperCase().trim();
}

function isFinishedStatus(status) {
  return ["FT", "AET", "PEN", "CANC", "ABD", "PST", "SUSP", "WO", "AWD"].includes(status);
}

// ✅ “sanidade”: se passou muito tempo do kickoff, derruba ao vivo
function isStaleLive(tsSec, elapsed) {
  if (!tsSec) return false;
  if (typeof elapsed !== "number" || elapsed < 80) return false;

  const kickoffMs = tsSec * 1000;
  const now = Date.now();
  const ageMin = (now - kickoffMs) / 60000;
  return ageMin >= LIVE_STALE_MINUTES;
}

function mapFixtureToGame(fx) {
  const fixture = fx.fixture || {};
  const teams = fx.teams || {};
  const league = fx.league || {};
  const goals = fx.goals || {};

  const id = fixture.id;
  const ts = fixture.timestamp; // segundos UTC

  const statusShort = fixture.status?.short;
  let status = normalizeStatusShort(statusShort);

  const longTxt = String(fixture.status?.long || "").toLowerCase();
  const elapsed = fixture.status?.elapsed ?? null;

  // live “normal”
  let isLive =
    ["1H", "2H", "HT", "ET", "BT", "P"].includes(status) ||
    longTxt.includes("live") ||
    longTxt.includes("in play") ||
    longTxt.includes("inplay");

  // ✅ se status é final, nunca ao vivo
  if (isFinishedStatus(status) || longTxt.includes("finished")) {
    isLive = false;
  }

  // ✅ se está “90’ preso” e já passou tempo demais, derruba ao vivo e marca como FT
  if (isLive && ts && isStaleLive(ts, elapsed)) {
    isLive = false;
    // se a API ainda não mandou FT, a gente “força” um status coerente
    status = "FT";
  }

  const brt = ts ? formatBRTFromTimestampSec(ts) : { date: null, time: null };

  const homeName = teams.home?.name || "Home";
  const awayName = teams.away?.name || "Away";

  const homeGoals = typeof goals.home === "number" ? goals.home : null;
  const awayGoals = typeof goals.away === "number" ? goals.away : null;

  return {
    id,
    home: homeName,
    away: awayName,
    league: league.name || "",
    country: league.country || "",
    date: brt.date || "",
    time: brt.time || "",
    kickoffTs: ts ? ts * 1000 : null, // ms
    status,
    isLive: !!isLive,
    elapsed,
    scoreHome: homeGoals,
    scoreAway: awayGoals,
  };
}

function uniqById(games) {
  const map = new Map();
  for (const g of games) {
    if (!g?.id) continue;

    const prev = map.get(g.id);
    if (!prev) {
      map.set(g.id, g);
      continue;
    }

    const prevScore = prev.scoreHome !== null || prev.scoreAway !== null;
    const newScore = g.scoreHome !== null || g.scoreAway !== null;

    if (newScore && !prevScore) map.set(g.id, g);
    else if (g.isLive && !prev.isLive) map.set(g.id, g);
    else map.set(g.id, { ...prev, ...g });
  }
  return [...map.values()];
}

async function main() {
  const { date: todayBRT } = brtNowParts();
  const tomorrowBRT = addDaysYYYYMMDD(todayBRT, 1);

  const [todayData, tomorrowData] = await Promise.all([
    apiGet("/fixtures", { date: todayBRT }),
    apiGet("/fixtures", { date: tomorrowBRT }),
  ]);

  let liveData = { response: [] };
  try {
    liveData = await apiGet("/fixtures", { live: "all" });
  } catch (e) {
    console.warn("⚠️ live=all falhou (ok no FREE às vezes). Seguindo sem live extra.");
  }

  const raw = [
    ...(todayData?.response || []),
    ...(tomorrowData?.response || []),
    ...(liveData?.response || []),
  ];

  const mapped = raw.map(mapFixtureToGame);

  const games = uniqById(mapped)
    .filter((g) => g.kickoffTs)
    .sort((a, b) => a.kickoffTs - b.kickoffTs);

  // ✅ updatedAt com segundos (pra você ver que está realmente atualizando)
  const updatedAt = new Intl.DateTimeFormat("pt-BR", {
    timeZone: TZ,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    hour: "2-digit",
    minute: "2-digit",
    second: "2-digit",
    hour12: false,
  }).format(new Date());

  const out = {
    updatedAt: `${updatedAt} BRT`,
    source: "API-Football",
    games,
  };

  const outPath = path.join(process.cwd(), "jogos.json");
  fs.writeFileSync(outPath, JSON.stringify(out, null, 2), "utf-8");
  console.log(`✅ jogos.json atualizado: ${games.length} jogos (hoje+amanhã+live).`);
}

main().catch((e) => {
  console.error("❌ Falha ao gerar jogos.json:", e?.message || e);
  process.exit(1);
});
